File: src/player.dart
import 'ship.dart';
import 'weapons.dart';

class Player extends Ship {
  static const MAX_HEALTH = 100;
  static const MAX_SPEED = 0.2;

  Player()
   : super(
     radius: 10, 
     mass: 1,
     health: MAX_HEALTH, 
     damage: 50, 
     speed: MAX_SPEED, 
     turnSpeed: 0.005, 
     color: 'green'
  ) {
    const GUN_ANGLE = 0.01;
    var playerShoot = () => new Bullet(damage: 10, color: color);

    // Left gun
    guns.add(new Gun(
      frontOffset: radius,
      sideOffset: -radius,
      angleOffset: GUN_ANGLE,
      speed: 0.4,
      timeBetweenShots: 100, 
      shoot: playerShoot, 
      owner: this
    ));

    // Right gun
    guns.add(new Gun(
      frontOffset: radius,
      sideOffset: radius,
      angleOffset: -GUN_ANGLE,
      speed: 0.4,
      timeBetweenShots: 100, 
      shoot: playerShoot, 
      owner: this
    ));

    // Missle launcher
    guns.add(new Gun(
      frontOffset: radius * 2,
      timeBetweenShots: 1000, 
      shoot: () => new Missle(damage: 50, speed: 0.3), 
      owner: this
    ));
  }
}

File: player.js
import { Ship } from "./ship.js"
import { Gun } from "./weapons.js"

class PlayerGun extends Gun {
   constructor(frontOffset, sideOffset, angleOffset, owner) {
      super({
         frontOffset: frontOffset,
         sideOffset: sideOffset,
         angleOffset: angleOffset,
         timeBetweenShots: 100,
         bulletSpeed: 0.4,
         bulletDamage: 10,
         bulletColor: Player.COLOR,
         owner: owner
      })
   }
}

export class Player extends Ship {
   static COLOR = "green"
   static MAX_HEALTH = 100

   constructor(x, y) {
      super({
         x: x, y: y, 
         radius: 10, 
         mass: 1,
         health: Player.MAX_HEALTH, 
         damage: 50, 
         speed: 0.2, 
         turnSpeed: 0.005,
         color: Player.COLOR
      })

      const GUN_ANGLE = 0.01
      const leftGun = new PlayerGun(this.radius, -this.radius, GUN_ANGLE, this)
      const rightGun = new PlayerGun(this.radius, this.radius, -GUN_ANGLE, this)

      this.setGuns(leftGun, rightGun)
   }

   update(dt) {
      this.turnTowardPoint(this.goalX, this.goalY, dt)

      super.update(dt)
   }
}